# Исключения

```C++
try
{
  ...
  if(error)
    throw 4;
  ...
}
catch(int e)
{
  ...
}
catch(int *e)
{
  ...
  throw e;
}
```

## Использование исключений для new

```C++
try
{
  S *p = new S;
}
catch(...) // ловить все ошибки
{
  printf("error");
  throw;
}
```

В 64 битных системах try-catch почти бесплатный, а в 32 битных требуется ресурсов

```C++
uint div(uint a, uint b)
{
  if (b == 0)
    throw "division by zero";
  return a / b;
}
```

# RAII

```C++
int *x = new int[a];
int *y = new int[b]; // Данный код приводит к утечке памяти
```
Вариант исправления
```C++
struct C
{
  int *p;
  S(int *z) : p(z) {}
  ~S() {delete[] p;}
}

S x(new int[a]);
S y(new int[b]);
```

Или

```C++
struct C
{
  int *p;
  S(size_t s) : p(new int [s]) {}
  ~S() {delete[] p;}
}

S x(a);
S y(b);
```
